File: vane-portals/src/main/java/org/oddlama/vane/portals/portal/PortalBlockLookup.java
package org.oddlama.vane.portals.portal;

import static org.oddlama.vane.core.persistent.PersistentSerializer.fromJson;
import static org.oddlama.vane.core.persistent.PersistentSerializer.toJson;

import java.io.IOException;
import java.util.UUID;
import org.jetbrains.annotations.NotNull;
import org.json.JSONObject;
import org.oddlama.vane.core.persistent.PersistentSerializer;

public class PortalBlockLookup {

    public static Object serialize(@NotNull final Object o) throws IOException {
        final var lookup = (PortalBlockLookup) o;
        final var json = new JSONObject();
        json.put("portalId", PersistentSerializer.toJson(UUID.class, lookup.portalId));
        json.put("type", PersistentSerializer.toJson(PortalBlock.Type.class, lookup.type));
        return json;
    }

    public static PortalBlockLookup deserialize(@NotNull final Object o) throws IOException {
        final var json = (JSONObject) o;
        final var portalId = PersistentSerializer.fromJson(UUID.class, json.get("portalId"));
        final var type = PersistentSerializer.fromJson(PortalBlock.Type.class, json.get("type"));
        return new PortalBlockLookup(portalId, type);
    }

    private UUID portalId;
    private PortalBlock.Type type;

    public PortalBlockLookup(final UUID portalId, final PortalBlock.Type type) {
        this.portalId = portalId;
        this.type = type;
    }

    public UUID portalId() {
        return portalId;
    }

    public PortalBlock.Type type() {
        return type;
    }
}

File: vane-portals/src/main/kotlin/org/oddlama/vane/portals/portal/PortalBlockLookup.kt
package org.oddlama.vane.portals.portal

import org.json.JSONObject
import org.oddlama.vane.core.persistent.PersistentSerializer
import java.io.IOException
import java.util.*

/**
 * Lightweight lookup payload that maps a block to a portal id and block type.
 *
 * @property portalId id of the owning portal.
 * @property type semantic type of the portal block.
 */
class PortalBlockLookup(private val portalId: UUID?, private val type: PortalBlock.Type?) {
    /** Returns the owning portal id. */
    fun portalId() = portalId

    /** Returns the semantic block type. */
    fun type() = type

    /** JSON serializer helpers for [PortalBlockLookup]. */
    companion object {
        /** Serializes a portal block lookup into a JSON object. */
        @JvmStatic
        @Throws(IOException::class)
        fun serialize(o: Any): Any {
            val lookup = o as PortalBlockLookup
            val json = JSONObject()
            json.put("portalId", PersistentSerializer.toJson(UUID::class.java, lookup.portalId))
            json.put("type", PersistentSerializer.toJson(PortalBlock.Type::class.java, lookup.type))
            return json
        }

        /** Deserializes a portal block lookup from a JSON object. */
        @JvmStatic
        @Throws(IOException::class)
        fun deserialize(o: Any): PortalBlockLookup {
            val json = o as JSONObject
            val portalId: UUID? = PersistentSerializer.fromJson(UUID::class.java, json.get("portalId"))
            val type: PortalBlock.Type? =
                PersistentSerializer.fromJson(PortalBlock.Type::class.java, json.get("type"))
            return PortalBlockLookup(portalId, type)
        }
    }
}
